File: backend/routers/billing.py
import logging

from fastapi import APIRouter, Depends, HTTPException, Request
from polar_sdk import Polar
from polar_sdk.webhooks import validate_event, WebhookVerificationError
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from config import settings
from database import get_db
from models import User
from services.auth import get_current_user

router = APIRouter(prefix="/billing", tags=["billing"])
logger = logging.getLogger(__name__)

# Initialize Polar client
polar = Polar(
    access_token=settings.polar_access_token,
    server=settings.polar_server,  # type: ignore[arg-type]
)


# Response Models


class CheckoutSessionResponse(BaseModel):
    checkout_url: str


# Polar API helper


async def cancel_polar_subscription(subscription_id: str) -> bool:
    """Cancel a subscription via Polar API (at period end)."""
    if not subscription_id:
        return False

    try:
        # Use update with cancel_at_period_end to allow access until period ends
        polar.subscriptions.update(
            id=subscription_id,
            subscription_update={"cancel_at_period_end": True},
        )
        logger.info(f"Canceled Polar subscription {subscription_id} (at period end)")
        return True
    except Exception:
        logger.exception(f"Error canceling subscription {subscription_id}")
        return False


# Endpoints


@router.post("/checkout-session", response_model=CheckoutSessionResponse)
async def create_checkout_session(user: User = Depends(get_current_user)):
    """Create a Polar checkout session for Pro subscription."""
    if user.plan == "pro":
        raise HTTPException(status_code=400, detail="Already subscribed to Pro")

    try:
        checkout = polar.checkouts.create(
            request={
                "products": [settings.polar_product_id],
                "success_url": f"{settings.frontend_url}/account?checkout=success&checkout_id={{CHECKOUT_ID}}",
                "customer_email": user.email,
                "embed_origin": settings.frontend_url,
                "metadata": {"user_id": user.id},
            }
        )
        return CheckoutSessionResponse(checkout_url=checkout.url)
    except Exception as e:
        logger.error(f"Failed to create checkout session: {e}")
        raise HTTPException(status_code=500, detail="Failed to create checkout session")


@router.post("/cancel-subscription")
async def cancel_subscription(user: User = Depends(get_current_user)):
    """Cancel the user's Pro subscription."""
    if user.plan != "pro":
        raise HTTPException(status_code=400, detail="No active subscription to cancel")

    if not user.polar_subscription_id:
        raise HTTPException(status_code=400, detail="No subscription found")

    success = await cancel_polar_subscription(user.polar_subscription_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to cancel subscription")

    return {"status": "ok"}


@router.post("/resubscribe")
async def resubscribe(user: User = Depends(get_current_user)):
    """Resubscribe by uncanceling a pending cancellation."""
    if not user.subscription_cancel_at_period_end:
        raise HTTPException(status_code=400, detail="No pending cancellation to undo")

    if not user.polar_subscription_id:
        raise HTTPException(status_code=400, detail="No subscription found")

    try:
        # Uncancel the subscription
        polar.subscriptions.update(
            id=user.polar_subscription_id,
            subscription_update={"cancel_at_period_end": False},
        )
        logger.info(f"Resubscribed user {user.email}")
        return {"status": "ok"}
    except Exception:
        logger.exception(f"Error resubscribing user {user.email}")
        raise HTTPException(status_code=500, detail="Failed to resubscribe")


@router.post("/webhook")
async def polar_webhook(request: Request, db: AsyncSession = Depends(get_db)):
    """Handle Polar webhook events."""
    payload = await request.body()
    headers = dict(request.headers)

    try:
        event = validate_event(payload, headers, settings.polar_webhook_secret)
    except WebhookVerificationError as e:
        logger.error(f"Webhook verification failed: {e}")
        raise HTTPException(status_code=400, detail="Invalid signature")

    event_type = event.TYPE
    subscription = event.data

    logger.info(f"Received Polar webhook: {event_type}")

    if event_type == "subscription.active":
        # Payment succeeded, subscription is now active
        await handle_subscription_active(subscription, db)
    elif event_type == "subscription.canceled":
        # User canceled, but subscription remains active until period ends
        await handle_subscription_canceled(subscription, db)
    elif event_type == "subscription.uncanceled":
        # User resubscribed before period ended
        await handle_subscription_uncanceled(subscription, db)
    elif event_type == "subscription.revoked":
        # Subscription has ended (after cancellation period or immediate revoke)
        await handle_subscription_revoked(subscription, db)
    elif event_type == "subscription.past_due":
        # Payment failed
        await handle_subscription_past_due(subscription, db)
    elif event_type == "subscription.updated":
        # General updates (renewal, plan change, etc.)
        await handle_subscription_updated(subscription, db)
    # Ignore: subscription.created (wait for active), checkout.*, etc.

    return {"status": "ok"}


# Webhook Handlers


async def handle_subscription_active(subscription, db: AsyncSession):
    """Handle subscription.active - payment succeeded, activate Pro plan."""
    customer_id = subscription.customer_id
    subscription_id = subscription.id
    customer_email = subscription.customer.email if subscription.customer else None

    if not customer_email:
        # Try to get email from metadata
        metadata = subscription.metadata or {}
        user_id = metadata.get("user_id")
        if user_id:
            result = await db.execute(select(User).where(User.id == user_id))
            user = result.scalar_one_or_none()
            if user:
                customer_email = user.email

    if not customer_email:
        logger.warning(f"No email found in subscription.active for {subscription_id}")
        return

    result = await db.execute(select(User).where(User.email == customer_email))
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for email {customer_email}")
        return

    user.plan = "pro"
    user.polar_customer_id = customer_id
    user.polar_subscription_id = subscription_id
    user.subscription_cancel_at_period_end = False  # Fresh subscription

    if subscription.current_period_end:
        user.plan_expires_at = subscription.current_period_end

    await db.commit()
    logger.info(f"Activated Pro for user {customer_email}")


async def handle_subscription_updated(subscription, db: AsyncSession):
    """Handle subscription.updated - renewal, plan change, etc."""
    subscription_id = subscription.id

    result = await db.execute(
        select(User).where(User.polar_subscription_id == subscription_id)
    )
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for subscription {subscription_id}")
        return

    # Update expiration date on renewal
    if subscription.current_period_end:
        user.plan_expires_at = subscription.current_period_end

    await db.commit()
    logger.info(f"Updated subscription for user {user.email}")


async def handle_subscription_canceled(subscription, db: AsyncSession):
    """Handle subscription.canceled - user canceled but still has access until period ends."""
    subscription_id = subscription.id

    result = await db.execute(
        select(User).where(User.polar_subscription_id == subscription_id)
    )
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for canceled subscription {subscription_id}")
        return

    # Mark as canceling - user keeps Pro until period ends
    user.subscription_cancel_at_period_end = True
    if subscription.current_period_end:
        user.plan_expires_at = subscription.current_period_end

    await db.commit()
    logger.info(
        f"Subscription canceled for user {user.email}, "
        f"access until {subscription.current_period_end}"
    )


async def handle_subscription_uncanceled(subscription, db: AsyncSession):
    """Handle subscription.uncanceled - user resubscribed before period ended."""
    subscription_id = subscription.id

    result = await db.execute(
        select(User).where(User.polar_subscription_id == subscription_id)
    )
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for uncanceled subscription {subscription_id}")
        return

    # User resubscribed - clear the cancel flag
    user.subscription_cancel_at_period_end = False

    await db.commit()
    logger.info(f"Subscription reactivated for user {user.email}")


async def handle_subscription_revoked(subscription, db: AsyncSession):
    """Handle subscription.revoked - subscription has ended, remove Pro access."""
    subscription_id = subscription.id

    result = await db.execute(
        select(User).where(User.polar_subscription_id == subscription_id)
    )
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for revoked subscription {subscription_id}")
        return

    user.plan = "free"
    user.polar_subscription_id = None
    user.plan_expires_at = None
    user.subscription_cancel_at_period_end = False
    # Keep polar_customer_id for potential resubscription

    await db.commit()
    logger.info(f"Subscription ended for user {user.email}, reverted to free plan")


async def handle_subscription_past_due(subscription, db: AsyncSession):
    """Handle subscription.past_due - payment failed."""
    subscription_id = subscription.id

    result = await db.execute(
        select(User).where(User.polar_subscription_id == subscription_id)
    )
    user = result.scalar_one_or_none()

    if not user:
        logger.warning(f"User not found for past_due subscription {subscription_id}")
        return

    # Log the issue - Polar will retry payment and send revoked if it fails permanently
    logger.warning(
        f"Payment failed for user {user.email}, subscription {subscription_id}. "
        "Polar will retry payment automatically."
    )

File: backend/src/routes/billing.rs
//! Billing endpoints — Polar checkout, subscription management, and webhooks.

use axum::extract::State;
use axum::http::{HeaderMap, StatusCode};
use axum::response::{IntoResponse, Response};
use axum::Json;
use base64::Engine;
use chrono::NaiveDateTime;
use hmac::{Mac, Hmac};
use serde::Deserialize;
use serde_json::{json, Value};
use sha2::Sha256;

use crate::auth::middleware::AuthUser;
use crate::error::error_response;
use crate::AppState;

/// Verify the Polar webhook signature using the standard-webhooks algorithm.
///
/// 1. Extract `webhook-id`, `webhook-timestamp`, `webhook-signature` headers.
/// 2. Verify the timestamp is within 5 minutes of now.
/// 3. Compute HMAC-SHA256 over `"{msg_id}.{timestamp}.{body}"`.
/// 4. Check if any `v1,` prefixed signature in the header matches.
fn verify_webhook_signature(
    headers: &HeaderMap,
    body: &[u8],
    secret: &str,
) -> Result<(), Response> {
    let msg_id = headers
        .get("webhook-id")
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| error_response(StatusCode::BAD_REQUEST, "Missing webhook-id header"))?;

    let timestamp_str = headers
        .get("webhook-timestamp")
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| {
            error_response(StatusCode::BAD_REQUEST, "Missing webhook-timestamp header")
        })?;

    let signatures = headers
        .get("webhook-signature")
        .and_then(|v| v.to_str().ok())
        .ok_or_else(|| {
            error_response(
                StatusCode::BAD_REQUEST,
                "Missing webhook-signature header",
            )
        })?;

    // Verify timestamp is within 5 minutes
    let timestamp: i64 = timestamp_str.parse().map_err(|_| {
        error_response(StatusCode::BAD_REQUEST, "Invalid webhook-timestamp")
    })?;
    let now = chrono::Utc::now().timestamp();
    if (now - timestamp).abs() > 300 {
        return Err(error_response(
            StatusCode::BAD_REQUEST,
            "Webhook timestamp too old or too far in the future",
        ));
    }

    // Build the signed content: "{msg_id}.{timestamp}.{body}"
    let body_str = std::str::from_utf8(body)
        .map_err(|_| error_response(StatusCode::BAD_REQUEST, "Invalid body encoding"))?;
    let to_sign = format!("{msg_id}.{timestamp_str}.{body_str}");

    // The net HMAC key is just secret.as_bytes() — Polar's SDK does
    // base64_encode(secret) then Webhook does base64_decode, which cancels out.
    let mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes())
        .map_err(|_| error_response(StatusCode::INTERNAL_SERVER_ERROR, "HMAC key error"))?;

    // The header may contain multiple space-separated signatures.
    // Use Mac::verify_slice for constant-time comparison.
    let valid = signatures.split(' ').any(|sig| {
        if let Some(b64) = sig.strip_prefix("v1,") {
            if let Ok(sig_bytes) = base64::engine::general_purpose::STANDARD.decode(b64) {
                let mut mac = mac.clone();
                mac.update(to_sign.as_bytes());
                return mac.verify_slice(&sig_bytes).is_ok();
            }
        }
        false
    });

    if !valid {
        return Err(error_response(StatusCode::BAD_REQUEST, "Invalid signature"));
    }

    Ok(())
}

/// Resolve the Polar API base URL from the config `polar_server` field.
fn polar_api_base(polar_server: &str) -> &str {
    if polar_server == "sandbox" {
        "https://sandbox-api.polar.sh"
    } else {
        "https://api.polar.sh"
    }
}

// ---------------------------------------------------------------------------
// Request / response models
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
struct WebhookPayload {
    #[serde(rename = "type")]
    event_type: String,
    data: Value,
}

// ---------------------------------------------------------------------------
// Endpoints
// ---------------------------------------------------------------------------

/// POST /billing/checkout-session — create a Polar checkout session (auth required).
pub async fn checkout_session(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
) -> Result<impl IntoResponse, Response> {
    if user.plan == "pro" {
        return Err(error_response(
            StatusCode::BAD_REQUEST,
            "Already subscribed to Pro",
        ));
    }

    let config = &state.config;

    // Test mode shortcut
    if config.test_mode {
        return Ok(Json(
            json!({"checkout_url": "https://test.polar.sh/checkout"}),
        ));
    }

    // Call Polar API to create a checkout session
    let base = polar_api_base(&config.polar_server);
    let client = &state.http_client;
    let resp = client
        .post(format!("{base}/v1/checkouts"))
        .header("Authorization", format!("Bearer {}", config.polar_access_token))
        .json(&json!({
            "product_price_id": config.polar_product_id,
            "success_url": format!("{}/checkout/success", config.frontend_url),
            "customer_email": user.email,
            "metadata": {
                "user_id": user.id,
            },
        }))
        .send()
        .await
        .map_err(|e| {
            tracing::error!("Polar checkout request failed: {e}");
            error_response(StatusCode::BAD_GATEWAY, "Failed to create checkout session")
        })?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        tracing::error!("Polar checkout error {status}: {body}");
        return Err(error_response(
            StatusCode::BAD_GATEWAY,
            "Polar API returned an error",
        ));
    }

    let data: Value = resp.json().await.map_err(|_| {
        error_response(StatusCode::BAD_GATEWAY, "Invalid response from Polar")
    })?;

    let checkout_url = data["url"]
        .as_str()
        .unwrap_or_default()
        .to_string();

    Ok(Json(json!({"checkout_url": checkout_url})))
}

/// POST /billing/cancel-subscription — cancel at period end (auth required).
pub async fn cancel_subscription(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
) -> Result<impl IntoResponse, Response> {
    if user.plan != "pro" {
        return Err(error_response(
            StatusCode::BAD_REQUEST,
            "No active Pro subscription",
        ));
    }

    let sub_id = user.polar_subscription_id.as_deref().ok_or_else(|| {
        error_response(StatusCode::BAD_REQUEST, "No subscription ID on record")
    })?;

    let config = &state.config;

    if config.test_mode {
        return Ok(Json(json!({"status": "ok"})));
    }

    let base = polar_api_base(&config.polar_server);
    let client = &state.http_client;
    let resp = client
        .patch(format!("{base}/v1/subscriptions/{sub_id}"))
        .header("Authorization", format!("Bearer {}", config.polar_access_token))
        .json(&json!({"cancel_at_period_end": true}))
        .send()
        .await
        .map_err(|e| {
            tracing::error!("Polar cancel request failed: {e}");
            error_response(StatusCode::BAD_GATEWAY, "Failed to cancel subscription")
        })?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        tracing::error!("Polar cancel error {status}: {body}");
        return Err(error_response(
            StatusCode::BAD_GATEWAY,
            "Polar API returned an error",
        ));
    }

    Ok(Json(json!({"status": "ok"})))
}

/// POST /billing/resubscribe — undo pending cancellation (auth required).
pub async fn resubscribe(
    State(state): State<AppState>,
    AuthUser(user): AuthUser,
) -> Result<impl IntoResponse, Response> {
    if !user.subscription_cancel_at_period_end {
        return Err(error_response(
            StatusCode::BAD_REQUEST,
            "Subscription is not pending cancellation",
        ));
    }

    let sub_id = user.polar_subscription_id.as_deref().ok_or_else(|| {
        error_response(StatusCode::BAD_REQUEST, "No subscription ID on record")
    })?;

    let config = &state.config;

    if config.test_mode {
        return Ok(Json(json!({"status": "ok"})));
    }

    let base = polar_api_base(&config.polar_server);
    let client = &state.http_client;
    let resp = client
        .patch(format!("{base}/v1/subscriptions/{sub_id}"))
        .header("Authorization", format!("Bearer {}", config.polar_access_token))
        .json(&json!({"cancel_at_period_end": false}))
        .send()
        .await
        .map_err(|e| {
            tracing::error!("Polar resubscribe request failed: {e}");
            error_response(StatusCode::BAD_GATEWAY, "Failed to resubscribe")
        })?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        tracing::error!("Polar resubscribe error {status}: {body}");
        return Err(error_response(
            StatusCode::BAD_GATEWAY,
            "Polar API returned an error",
        ));
    }

    Ok(Json(json!({"status": "ok"})))
}

/// POST /billing/webhook — Polar webhook receiver (NO auth).
pub async fn webhook(
    State(state): State<AppState>,
    headers: HeaderMap,
    body: axum::body::Bytes,
) -> Result<impl IntoResponse, Response> {
    let config = &state.config;

    // Verify webhook signature
    verify_webhook_signature(&headers, &body, &config.polar_webhook_secret)?;

    // Parse payload
    let payload: WebhookPayload = serde_json::from_slice(&body).map_err(|e| {
        tracing::error!("Failed to parse webhook body: {e}");
        error_response(StatusCode::BAD_REQUEST, "Invalid JSON body")
    })?;

    let data = &payload.data;
    let db = &state.db;

    match payload.event_type.as_str() {
        "subscription.active" => {
            // Find user by customer email, or metadata.user_id fallback
            let email = data
                .pointer("/customer/email")
                .and_then(|v| v.as_str());
            let metadata_user_id = data
                .pointer("/metadata/user_id")
                .and_then(|v| v.as_str());

            let user_id: Option<String> = if let Some(email) = email {
                sqlx::query_scalar("SELECT id FROM users WHERE email = ?")
                    .bind(email)
                    .fetch_optional(db)
                    .await
                    .unwrap_or(None)
            } else {
                None
            };

            let user_id = user_id.or_else(|| metadata_user_id.map(|s| s.to_string()));

            let Some(user_id) = user_id else {
                tracing::warn!(
                    "subscription.active: could not resolve user (email={email:?}, metadata_user_id={metadata_user_id:?})"
                );
                return Ok(Json(json!({"status": "ok"})));
            };

            let customer_id = data["customer_id"].as_str().filter(|s| !s.is_empty());
            let subscription_id = data["id"].as_str().unwrap_or_default();
            let period_end = parse_period_end(data.get("current_period_end"));

            sqlx::query(
                "UPDATE users SET plan = 'pro', polar_customer_id = COALESCE(?, polar_customer_id),
                 polar_subscription_id = ?,
                 subscription_cancel_at_period_end = 0, plan_expires_at = ? WHERE id = ?",
            )
            .bind(customer_id)
            .bind(subscription_id)
            .bind(period_end)
            .bind(&user_id)
            .execute(db)
            .await
            .map_err(|e| {
                tracing::error!("DB error on subscription.active: {e}");
                error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error")
            })?;

            tracing::info!("subscription.active: user {user_id} upgraded to pro");
        }

        "subscription.canceled" => {
            let sub_id = data["id"].as_str().unwrap_or_default();
            let period_end = parse_period_end(data.get("current_period_end"));

            sqlx::query(
                "UPDATE users SET subscription_cancel_at_period_end = 1, plan_expires_at = ?
                 WHERE polar_subscription_id = ?",
            )
            .bind(period_end)
            .bind(sub_id)
            .execute(db)
            .await
            .map_err(|e| {
                tracing::error!("DB error on subscription.canceled: {e}");
                error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error")
            })?;

            tracing::info!("subscription.canceled: sub {sub_id} marked cancel-at-period-end");
        }

        "subscription.uncanceled" => {
            let sub_id = data["id"].as_str().unwrap_or_default();

            sqlx::query(
                "UPDATE users SET subscription_cancel_at_period_end = 0
                 WHERE polar_subscription_id = ?",
            )
            .bind(sub_id)
            .execute(db)
            .await
            .map_err(|e| {
                tracing::error!("DB error on subscription.uncanceled: {e}");
                error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error")
            })?;

            tracing::info!("subscription.uncanceled: sub {sub_id} reactivated");
        }

        "subscription.revoked" => {
            let sub_id = data["id"].as_str().unwrap_or_default();

            sqlx::query(
                "UPDATE users SET plan = 'free', polar_subscription_id = NULL,
                 plan_expires_at = NULL, subscription_cancel_at_period_end = 0
                 WHERE polar_subscription_id = ?",
            )
            .bind(sub_id)
            .execute(db)
            .await
            .map_err(|e| {
                tracing::error!("DB error on subscription.revoked: {e}");
                error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error")
            })?;

            tracing::info!("subscription.revoked: sub {sub_id} downgraded to free");
        }

        "subscription.updated" => {
            let sub_id = data["id"].as_str().unwrap_or_default();
            let period_end = parse_period_end(data.get("current_period_end"));

            sqlx::query(
                "UPDATE users SET plan_expires_at = ? WHERE polar_subscription_id = ?",
            )
            .bind(period_end)
            .bind(sub_id)
            .execute(db)
            .await
            .map_err(|e| {
                tracing::error!("DB error on subscription.updated: {e}");
                error_response(StatusCode::INTERNAL_SERVER_ERROR, "Database error")
            })?;

            tracing::info!("subscription.updated: sub {sub_id} period end updated");
        }

        "subscription.past_due" => {
            tracing::warn!(
                "subscription.past_due: sub {} is past due",
                data["id"].as_str().unwrap_or("?")
            );
            // No DB changes — just log.
        }

        other => {
            tracing::debug!("Unhandled webhook event type: {other}");
        }
    }

    Ok(Json(json!({"status": "ok"})))
}

/// Parse `current_period_end` from the webhook data into a NaiveDateTime.
/// Accepts ISO 8601 strings (e.g. "2025-05-01T00:00:00Z").
fn parse_period_end(value: Option<&Value>) -> Option<NaiveDateTime> {
    value
        .and_then(|v| v.as_str())
        .and_then(|s| {
            // Try parsing as full DateTime first, then fall back to NaiveDateTime
            s.parse::<chrono::DateTime<chrono::Utc>>()
                .map(|dt| dt.naive_utc())
                .ok()
                .or_else(|| NaiveDateTime::parse_from_str(s, "%Y-%m-%dT%H:%M:%S").ok())
        })
}
